bool success
int64 blah